plugin.audio.ampache/default.py

```diff
@@ -0,0 +1,1070 @@
+def addSongLinks(elem):
+   
+    xbmcplugin.setContent(int(sys.argv[1]), "songs")
+    ok=True
+    it=[]
+    for node in elem.iter("song"):
+        liz=xbmcgui.ListItem()
+        fillListItemWithSongInfo(liz,node)
+        liz.setProperty("IsPlayable", "true")
+
+        cm = []
+        try:
+            artist_id = getSuperId(node,"artist")
+            cm.append( ( ut.tString(30138),
+            "Container.Update(%s?object_id=%s&mode=15)" % (
+                sys.argv[0],artist_id ) ) )
+        except:
+            pass
+        
+        try:
+            album_id = getSuperId(node,"album")
+            cm.append( ( ut.tString(30139),
+            "Container.Update(%s?object_id=%s&mode=16)" % (
+                sys.argv[0],album_id ) ) )
+        except:
+            pass
+        
+        song_title = str(node.findtext("title"))
+        cm.append( ( ut.tString(30140),
+        "Container.Update(%s?title=%s&mode=17)" % (
+            sys.argv[0],song_title) ) )
+
+        if cm != []:
+            liz.addContextMenuItems(cm)
+
+        song_url = node.findtext("url")
+        song_id = int(node.attrib["id"])
+        track_parameters = { "mode": 45, "song_url" : song_url, "object_id" : song_id}
+        url = sys.argv[0] + '?' + urllib.parse.urlencode(track_parameters)
+        tu= (url,liz)
+        it.append(tu)
+    
+    ok=xbmcplugin.addDirectoryItems(handle=int(sys.argv[1]),items=it,totalItems=len(elem))
+    #@xbmc.log("AmpachePlugin::addSongLinks " + str(ok), xbmc.LOGDEBUG)
+    return ok
+
+#The function that actually plays an Ampache URL by using setResolvedUrl ( via
+#AmpachePlayer class
+def play_track(object_id,song_url):
+    if song_url == None or object_id == None:
+        xbmc.log("AmpachePlugin::play_track object or song null", xbmc.LOGNOTICE )
+        return
+
+    old_object_id = None
+
+    #it checks if we need the song infolabels ( object_id cached is different from
+    #object_id of the song, for instance as kore app call the song
+    try:
+            plugin_url = xbmc.getInfoLabel('ListItem.FileNameAndPath')
+            params=ut.get_params(plugin_url)
+            old_object_id=int(params["object_id"])
+            xbmc.log("AmpachePlugin::play_track old_object_id " + str(old_object_id), xbmc.LOGDEBUG)
+    except:
+            pass
+
+    liz = xbmcgui.ListItem()
+    try:
+        if old_object_id == None or old_object_id != object_id:
+            ampConn = ampache_connect.AmpacheConnect()
+            xbmc.log("AmpachePlugin::play_track refresh infoLabels", xbmc.LOGDEBUG)
+            ampConn.filter = object_id
+            elem = ampConn.ampache_http_request("song")
+            for thisnode in elem:
+                node = thisnode
+            fillListItemWithSongInfo(liz,node)
+            liz.setProperty("IsPlayable", "true")
+    except:
+        pass
+
+    liz.setPath(song_url)
+    AmpachePlayer = player.AmpachePlayer()
```

Comment:
please use setResolvedUrl, plugins are not allowed to call Player.play directly

Reply:
AmpachePlayer is a class built in the plugin that use setResolvedUrl, as you say, i have added it to add some checks if it is necessary

Reply:
Ok, not perfect, but done